File: algorithms/Strings/Medium/BasicCalculator/VSDBasicCalculator.py


def calculator(s):
    stack=list()
    curnum=0
    op="+"
    for i in range(len(s)):
        if s[i].isdigit():
            curnum=curnum*10+int(s[i])
        if (not (s[i].isdigit()) and s[i]!=" ") or i==len(s)-1:
            if op=="-":
                stack.append(-curnum)
            elif op=="+":
                stack.append(curnum)
            elif op=="*":
                stack.append(stack.pop()*curnum)
            elif op=="/":
                x=stack.pop()
                if x<0:
                    stack.append(-(abs(x)//curnum))
                else:
                    stack.append(x//curnum)
            op=s[i]
            curnum=0
    ans=0
    while stack:
        ans+=stack.pop()
    return ans 


s = "3+2*2"
print(calculator(s))

File: algorithms/Strings/Medium/BasicCalculator/VSDBasicCalculator.cpp
#include <bits/stdc++.h>

using namespace std;

int calculator(string s)
{
    stack<int> stack;
    int curnum = 0;
    char op = '+';

    for (int i = 0; i < s.length(); i++)
    {

        if (s[i] >= 48 && s[i] <= 57)
            curnum = curnum * 10 + int(s[i] - 48);

        if (!(s[i] >= 48 && s[i] <= 57 && s[i] != ' ') || i == s.length() - 1)
        {

            if (op == '-')
                stack.push(-curnum);

            else if (op == '+')
                stack.push(curnum);

            else if (op == '*')
            {
                int x = stack.top();
                stack.pop();
                stack.push(curnum * x);
            }

            else if (op == '/')
            {
                int x = stack.top();
                stack.pop();
                if (x < 0)
                    stack.push(-(abs(x) / curnum));
                else
                    stack.push(x / curnum);
            }
            op = s[i];
            curnum = 0;
        }
    }
    int ans = 0;

    while (!stack.empty())
    {
        ans += stack.top();
        stack.pop();
    }
    return ans;
}

int main()
{

    string s = "3+2*2";
    cout << calculator(s);
    return 0;
}